File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int maxArea = INT_MIN;
        stack<int> indexes;
        for(int i = 0; i < heights.size(); i++){
            while(!indexes.empty() && heights[indexes.top()] > heights[i]){
                int minIdx = indexes.top();
                indexes.pop();

                int width = indexes.empty() ? i : i - indexes.top() - 1;
                maxArea = max(maxArea, heights[minIdx] * width);
            }
            indexes.push(i);
        }

        while(!indexes.empty()){
            int minIdx = indexes.top();
            indexes.pop();

            int width = indexes.empty() ? heights.size() : heights.size() - indexes.top() - 1;
            maxArea = max(maxArea, heights[minIdx] * width);
        }
        return maxArea;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        area = 0
        l = 0
        r = len(heights) - 1
        while l < r:
            minHeight = min(heights[l], heights[r])
            currArea = minHeight * (r - l)
            area = max(currArea, area)
            if heights[l] < heights[r]: l += 1
            else: r -= 1
        return area